File: app/src/types/game.ts
export enum GameState {
  NotStarted = 'NotStarted',
  Starting = 'Starting',
  RollDice = 'RollDice',
  RollingDice = 'RollingDice',
  Move = 'Move',
  Finished = 'Finished',
}

export enum Colors {
  Red = 0,
  Green = 1,
  Yellow = 2,
  Blue = 3,
}

export interface Game {
  seed: number;
  bump: number;
  numPlayers: number;
  curPlayer: number;
  tokenPositions: number[][]; // [player][token] -> position
  gameState: GameState;
  currentRoll: number;
  sixCount: number;
  players: string[]; // Public keys
  winner: string;
}

export interface Player {
  publicKey: string;
  color: Colors;
  isCurrentPlayer: boolean;
}

export interface GameAction {
  type: 'create' | 'join' | 'roll' | 'move' | 'tokenIntoPlay' | 'cancel';
  payload?: Record<string, unknown>;
}

export interface DiceRoll {
  value: number;
  isRolling: boolean;
}

export interface Token {
  id: number;
  position: number;
  color: Colors;
  isInPlay: boolean;
  isSelected: boolean;
} 

File: programs/ludo/src/state/game.rs
use anchor_lang::prelude::*;

#[account]
#[derive(InitSpace)]
pub struct Game {
    pub seed: u64,
    pub bump: u8,
    pub num_players: u8,
    pub cur_player: u8,
    pub token_positions: [[i8; 4]; 4],
    pub game_state: GameState,
    pub current_roll: u8,
    pub six_count: u8,
    pub players: [Pubkey; 4],
    pub winner: Pubkey, // ??? is there one winner or multiple???
}

impl Game {
    pub fn next_player(&mut self) {
        loop {
            self.cur_player = (self.cur_player + 1) % 4;
            if self.cur_player_key() != Pubkey::default() {
                break;
            }
        }
    }

    #[inline(always)]
    pub fn cur_player_key(&self) -> Pubkey {
        self.players[self.cur_player as usize]
    }
}

#[derive(AnchorSerialize, AnchorDeserialize, Clone, PartialEq, InitSpace)]
pub enum GameState {
    NotStarted,
    Starting,
    RollDice,
    RollingDice,
    Move,
    Finished,
}
